### ISSUE TITLE
problem when changing Windows default playback device during NAudio playback

### ISSUE THREAD

**Issue by yesyesuk** [naudio/NAudio#708]
Hi,
I'm new to NAudio. 
I have written a small WinForms app that periodically plays either a tone (using the SignalGenerator) or a wav.mp3 file using an audio playback device that I can select in the app. I want it on a specific device in order to prevent my external speakers from going to standby.

In general it works fine. When I start the app and it starts playing on my selected speakers (the actual audio device, not "Microsoft Sound Mapper"). However, when I then go into the Windows audio devices settings and change the default playback device, the playback changes to that new default audio device. This is exactly what I wanted to prevent by selecting a specific audio device.

What could be causing this? How do I prevent this from happening?
I want to be able to change the default playback device in Windows but my app should always play on the selected device.

**Comment by markheath** [comment]
What IWavePlayer are you using WaveOut? WasapiOut? something else?

**Comment by yesyesuk** [comment]
Hi. I use WaveOutEvent

`private WaveOutEvent outputDevice;
`
then before play I do:

`outputDevice = new WaveOutEvent() { DeviceNumber = GetSelectedDeviceNumber() };
`
GetSelectedDeviceNumber() does this:

```
        private int GetSelectedDeviceNumber()
        {
            for (int n = -1; n < WaveOut.DeviceCount; n++)
            {
                WaveOutCapabilities caps = WaveOut.GetCapabilities(n);
                if (audioDevice == caps.ProductName) return n;
            }
            // not found
            return -1;
        }

```
"audioDevice" is a string that contains the name of the selected audio device from a comboBox.


In the meantime I have found a workaround. Since I only play a tone or audio file periodically (for example for 10 seconds every 60 seconds), before each new play I dispose outputDevice and create a new one with the audio device I want to play on. This works fine for a few weeks now.

It would be much more problematic for continuous plays.

**Comment by markheath** [comment]
Are you sure you're not using -1 
-1 is actually a valid device number with waveOut and it means to use the mapper

But under the hood NAudio is just calling waveOutOpen with the device number you choose, so the behaviour you will get is just however waveOut works

**Comment by yesyesuk** [comment]
Hmmm, I never actually checked the number. But I'm quite sure it wasn't -1. I understand that device number -1is basically the Windows default playback device, whichever that is at the time. (please correct me if that's wrong).

However, what happened in my app was that I played on a specific device that was not the default at the time, and it did indeed play on that device. Then I went into the Windows audio settings and changed the default playback device. My app then started playing on that new default device. At that time the app only called
`outputDevice = new WaveOutEvent() { DeviceNumber = GetSelectedDeviceNumber() };`
on start-up of the app or when I manually selected a new audio device from my comboBox.

Could it be that the device numbers change when the Windows default playback device is being changed?

**Comment by markheath** [comment]
yes, the device numbers will change - they are just a zero-based count so won't remain constant if you are adding and removing usb audio devices for example

**Comment by markheath** [comment]
The cause is that `WaveOut`/`WaveOutEvent` `DeviceNumber` is an unstable **zero-based index**, not a stable reference to a physical device — and `ProductName` is truncated to 31 characters and isn't guaranteed unique. When you change the Windows default, the index→endpoint mapping shifts, so on the next play the same number resolves to a different device. That's inherent to the legacy WinMM API, not something NAudio can pin down through it.

NAudio 3 gives you a proper fix: use **`WasapiPlayer`**, which targets a device by its **stable endpoint ID** rather than a shifting index, so playback stays on the device you chose no matter what the default becomes:

```csharp
var enumerator = new MMDeviceEnumerator();
var device = enumerator.EnumerateAudioEndPoints(DataFlow.Render, DeviceState.Active)
                       .First(d => d.ID == savedEndpointId);   // match by ID, not name

using var player = new WasapiPlayerBuilder().WithDevice(device).Build();
player.Init(yourSource);
player.Play();
```

Save each device's `MMDevice.ID` (a stable string) rather than its name or index, and match on that when you open the player.

Conversely, if you ever want the *opposite* — playback that automatically follows the default device as it changes — NAudio 3 now supports that explicitly via `WasapiPlayerBuilder().WithDefaultDeviceStreamRouting()`. So both behaviours are now first-class and opt-in, instead of depending on WinMM index quirks.

Closing as resolved in NAudio 3 - the WASAPI player pinned to a specific endpoint does exactly what you were after.
